src/commonMain/kotlin/exchange.dydx.abacus/functional/vault/Vault.kt

```diff
@@ -147,49 +151,114 @@ object VaultCalculator {
 
         val historiesMap = histories?.vaultsPnl?.associateBy { it.ticker }
 
-        return VaultPositions(positions = positions.positions.mapNotNull { calculateVaultPosition(it, historiesMap?.get(it.ticker), markets?.get(it.ticker)) }.toIList())
+        return VaultPositions(
+            positions = positions.positions.mapNotNull {
+                calculateVaultPosition(
+                    it,
+                    historiesMap?.get(it.ticker),
+                    markets?.get(it.ticker),
+                )
+            }.toIList(),
+        )
     }
 
-    fun calculateVaultPosition(position: IndexerVaultPosition, history: IndexerVaultHistoricalPnl?, perpetualMarket: PerpetualMarket?): VaultPosition? {
-        if (position.ticker == null) {
+    internal fun calculateVaultPositionsInternal(
+        vault: InternalVaultState?,
+        markets: IMap<String, PerpetualMarket>?
+    ): VaultPositions? {
+        if (vault?.positions == null) {
             return null
         }
-        val perpetualPosition = perpetualPositionProcessor.process(null, position.perpetualPosition)
-        val assetPosition = assetPositionProcessor.process(position.assetPosition)
 
+        val positions: List<VaultPosition>? = vault.positions?.mapNotNull { position ->
+            val ticker = position.ticker ?: return@mapNotNull null
+            val history = vault.pnls.get(ticker)
+            val market = markets?.get(ticker)
+            calculateVaultPositionInternal(
+                ticker = ticker,
+                equity = position.equity,
+                perpetualPosition = position.openPosition,
+                assetPosition = position.assetPosition,
+                thirtyDayPnl = history,
+                perpetualMarket = market,
+            )
+        }
+        return VaultPositions(positions = positions?.toIList())
+    }
+
+    fun calculateVaultPosition(
+        position: IndexerVaultPosition,
+        history: IndexerVaultHistoricalPnl?,
+        perpetualMarket: PerpetualMarket?
+    ): VaultPosition? {
+        if (position.ticker != null) {
+            val perpetualPosition =
+                perpetualPositionProcessor.process(null, position.perpetualPosition)
+            val assetPosition = assetPositionProcessor.process(position.assetPosition)
+            val thirtyDayPnl = calculateThirtyDayPnl(history)
+            return calculateVaultPositionInternal(
+                ticker = position.ticker,
+                equity = parser.asDouble(position.equity),
+                perpetualPosition = perpetualPosition,
+                assetPosition = assetPosition,
+                thirtyDayPnl = thirtyDayPnl,
+                perpetualMarket = perpetualMarket,
+            )
+        } else {
+            return null
+        }
+    }
+
+    internal fun calculateVaultPositionInternal(
```

Comment:
Performance optimization.  I separated this out (and made `calculateThirtyDayPnl` internal) because we can process `perpetualPosition`, `assetPositionProcessor` and `thirtyDayPnl` when indexer data arrive.  @tyleroooo